File: python/src/agent_framework/core/session_lane.py
from __future__ import annotations

import asyncio


class DefaultSessionLane:
    def __init__(self) -> None:
        self._queues: dict[str, asyncio.Task | None] = {}

    async def enqueue(self, session_id: str, task):
        previous = self._queues.get(session_id)
        if previous is not None:
            try:
                await previous
            except Exception:
                pass

        current = asyncio.create_task(task())
        self._queues[session_id] = current
        try:
            return await current
        finally:
            if self._queues.get(session_id) is current:
                self._queues.pop(session_id, None)

File: typescript/packages/core/src/session-lane.ts
import type { SessionLane } from './types/contracts.js';

export class DefaultSessionLane implements SessionLane {
  private queues = new Map<string, Promise<unknown>>();

  enqueue<T>(sessionId: string, task: () => Promise<T>): Promise<T> {
    const previous = this.queues.get(sessionId) ?? Promise.resolve();
    const next = previous
      .catch(() => undefined)
      .then(task);
    this.queues.set(sessionId, next);
    return next.finally(() => {
      if (this.queues.get(sessionId) === next) {
        this.queues.delete(sessionId);
      }
    }) as Promise<T>;
  }
}
